edb/server/compiler/rpc.pyx

```diff
@@ -0,0 +1,290 @@
+
+    def update(
+        self,
+        source: edgeql.Source,
+        protocol_version: defines.ProtocolVersion,
+        *,
+        output_format: enums.OutputFormat = OUT_FMT_BINARY,
+        input_format: enums.InputFormat = IN_FMT_BINARY,
+        expect_one: bint = False,
+        implicit_limit: int = 0,
+        inline_typeids: bint = False,
+        inline_typenames: bint = False,
+        inline_objectids: bint = True,
+    ) -> CompileRequest:
+        self.source = source
+        self.protocol_version = protocol_version
+        self.output_format = output_format
+        self.json_parameters = input_format is IN_FMT_JSON
+        self.expect_one = expect_one
+        self.implicit_limit = implicit_limit
+        self.inline_typeids = inline_typeids
+        self.inline_typenames = inline_typenames
+        self.inline_objectids = inline_objectids
+
+        self.serialized_cache = None
+        self.cache_key = None
+        return self
+
+    def set_modaliases(self, value) -> CompileRequest:
+        self.modaliases = value
+        self.serialized_cache = None
+        self.cache_key = None
+        return self
+
+    def set_session_config(self, value) -> CompileRequest:
+        self.session_config = value
+        self.serialized_cache = None
+        self.cache_key = None
+        return self
+
+    def set_system_config(self, value) -> CompileRequest:
+        self.system_config = value
+        self.serialized_cache = None
+        self.cache_key = None
+        return self
+
+    def deserialize(self, bytes data, str query_text) -> CompileRequest:
+        if data[0] == 0:
+            self._deserialize_v0(data, query_text)
+        else:
+            raise errors.UnsupportedProtocolVersionError(
+                f"unsupported compile cache: version {data[0]}"
+            )
+        return self
+
+    def serialize(self) -> bytes:
+        if self.serialized_cache is None:
+            self._serialize()
+        return self.serialized_cache
+
+    def get_cache_key(self) -> uuid.UUID:
+        if self.cache_key is None:
+            self._serialize()
+        return self.cache_key
+
+    cdef _serialize(self):
+        cdef:
+            char version = 0, flags
+            WriteBuffer out = WriteBuffer.new()
+
+        out.write_byte(version)
+
+        flags = (
+            (MASK_JSON_PARAMETERS if self.json_parameters else 0) |
+            (MASK_EXPECT_ONE if self.expect_one else 0) |
+            (MASK_INLINE_TYPEIDS if self.inline_typeids else 0) |
+            (MASK_INLINE_TYPENAMES if self.inline_typenames else 0) |
+            (MASK_INLINE_OBJECTIDS if self.inline_objectids else 0)
+        )
+        out.write_byte(flags)
+
+        out.write_int16(self.protocol_version[0])
+        out.write_int16(self.protocol_version[1])
+        out.write_byte(serialize_output_format(self.output_format))
+        out.write_int64(self.implicit_limit)
+
+        if self.modaliases is None:
```

Comment:
> Is it important to distinguish the None vs empty cases?

Probably not, but I didn't verify so decided to distinguish for now.

> If so, maybe it would be better to do None as -1?

Ah, I was scared by the `cdef read_int32() except? -1`. Looks like `except?` is fine with returning `-1` but only adds a little bit of overhead:

> The `?` indicates that the value `-1` *may* signal an error.
>
> In this case, Cython generates a call to [PyErr_Occurred()](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Occurred) if the exception value is returned, to make sure it really received an exception and not just a normal result.

Reply:
:+1:

Maybe we should change it to `except? -2` or something, also!

Reply:
Because `except? -1` is in another repo, I'll use `-2` in the serialization for now, the deserialization is compatible (>= 0).

I'll create another PR for pgproto to change to `cdef read_intN() except? -1 << N - 2` as the most unlikely value.

Reply:
I would just use `-1` for now. No reason to write something weird out to disk for a temporary microoptimization.

Reply:
Not that it really matters, since the "spec" can just say negative means empty